File: src/main/java/com/dengzii/plugin/findview/MainAction.java
package com.dengzii.plugin.findview;

import com.dengzii.plugin.findview.utils.XLog;
import com.intellij.openapi.actionSystem.AnAction;
import com.intellij.openapi.actionSystem.AnActionEvent;
import com.intellij.openapi.actionSystem.LangDataKeys;
import com.intellij.openapi.actionSystem.PlatformDataKeys;
import com.intellij.openapi.editor.Document;
import com.intellij.openapi.editor.Editor;
import com.intellij.openapi.editor.event.DocumentEvent;
import com.intellij.openapi.editor.event.DocumentListener;
import com.intellij.openapi.project.Project;
import com.intellij.psi.*;
import com.intellij.psi.impl.file.PsiDirectoryFactory;
//import com.intellij.psi.impl.source.PsiJavaFileImpl;
import com.intellij.psi.util.PsiTreeUtil;
import org.jetbrains.annotations.NotNull;

import java.util.Arrays;

public class MainAction extends AnAction {

    private static final String TAG = "MainAction";

    private PsiClass mClass;
    private PsiJavaFile mPsiJavaFile;
    private Project mProject;
    private PsiFile mPsiFile;

//    public MainAction() {
//        super(null);
//    }
//
//    protected MainAction(CodeInsightActionHandler handler) {
//        super(handler);
//    }

    @Override
    public void actionPerformed(AnActionEvent e) {

        Editor editor = e.getData(PlatformDataKeys.EDITOR);

        this.mProject = e.getProject();
        this.mPsiFile = e.getData(LangDataKeys.PSI_FILE);
        this.mClass = getTargetClass(editor, mPsiFile);

        mPsiJavaFile = ((com.intellij.psi.impl.source.PsiJavaFileImpl) mPsiFile);

        System.out.println(mPsiJavaFile);

        Document document = editor.getDocument();
        XLog.info(TAG, "className: " + mClass.getName());
        XLog.info(TAG, "superClass: " + Arrays.toString(mClass.getSupers()));
        XLog.info(TAG, "fields: " + Arrays.toString(mClass.getAllFields()));

        document.addDocumentListener(new DocumentListener() {
            @Override
            public void beforeDocumentChange(@NotNull DocumentEvent event) {
                XLog.info(TAG, "beforeDocumentChange: " + event.getOldFragment());
            }

            @Override
            public void documentChanged(@NotNull DocumentEvent event) {
                XLog.info(TAG, "documentChanged: " + event.getNewFragment());
            }
        });
    }

    private PsiClass getTargetClass(Editor editor, PsiFile file) {
        int offset = editor.getCaretModel().getOffset();
        PsiElement element = file.findElementAt(offset);
        if (element == null) {
            return null;
        } else {
            PsiClass target = PsiTreeUtil.getParentOfType(element, PsiClass.class);
            return target instanceof SyntheticElement ? null : target;
        }
    }
}

File: src/main/kotlin/com/dengzii/plugin/findview/MainAction.kt
package com.dengzii.plugin.findview

import com.dengzii.plugin.findview.gen.FindViewCodeWriter
import com.dengzii.plugin.findview.gen.GenConfig
import com.dengzii.plugin.findview.tools.ui.showAtCenterScreen
import com.dengzii.plugin.findview.ui.XGenDialog
import com.dengzii.plugin.findview.utils.PsiFileUtils
import com.intellij.openapi.actionSystem.ActionUpdateThread
import com.intellij.openapi.actionSystem.AnAction
import com.intellij.openapi.actionSystem.AnActionEvent
import com.intellij.openapi.actionSystem.CommonDataKeys
import com.intellij.openapi.command.WriteCommandAction

/**
 * <pre>
 * author : dengzi
 * e-mail : dengzii@foxmail.com
 * github : [...](https://github.com/dengzii)
 * time   : 2019/9/27
 * desc   :
</pre> *
 */
class MainAction : AnAction() {
    override fun actionPerformed(e: AnActionEvent) {
        val project = e.project
        val psiFile = e.getData(CommonDataKeys.PSI_FILE)
        val editor = e.getData(CommonDataKeys.EDITOR)
        if (isNull(project!!, psiFile!!, editor!!)) {
            return
        }
        val vis = PsiFileUtils.getViewInfoFromPsiFile(psiFile, project)
        val conf = GenConfig(emptyList(), psiFile, editor)
        XGenDialog(conf, vis) {
            if (it.viewInfo.isEmpty()) {
                return@XGenDialog
            }
            WriteCommandAction.writeCommandAction(project).run(FindViewCodeWriter(it))
        }.showAtCenterScreen()
    }

    override fun getActionUpdateThread(): ActionUpdateThread {
        return ActionUpdateThread.BGT
    }

    override fun update(e: AnActionEvent) {
        super.update(e)
        val project = e.project
        val psiFile = e.getData(CommonDataKeys.PSI_FILE)
        val editor = e.getData(CommonDataKeys.EDITOR)
        e.presentation.isEnabledAndVisible = true
        if (isNull(project, psiFile, editor)) {
            e.presentation.isEnabledAndVisible = false
            return
        }
        val language = psiFile!!.language
        if (language != Config.JAVA && language != Config.KOTLIN) {
            e.presentation.isEnabledAndVisible = false
        }
    }

    private fun isNull(vararg objects: Any?): Boolean {
        for (o in objects) {
            if (o == null) {
                return true
            }
        }
        return false
    }
}
